homeassistant/helpers/intent.py

```diff
@@ -554,12 +602,25 @@ def async_match_targets(  # noqa: C901
         # Filter by entity name or alias
         candidates = list(_filter_by_name(constraints.name, candidates))
         if not candidates:
+            # Check if any unexposed entities match the name
+            for state in unexposed_states:
+                if _matches_name(state, constraints.name, er):
+                    return MatchTargetsResult(False, MatchFailedReason.ASSISTANT)
+
             return MatchTargetsResult(False, MatchFailedReason.NAME)
 
     if constraints.features:
         # Filter by supported features
         candidates = list(_filter_by_features(constraints.features, candidates))
         if not candidates:
+            # Check if any unexposed entities match the supported features
+            for state in unexposed_states:
+                if (
+                    entity_helper.get_supported_features(hass, state.entity_id)
+                    == constraints.features
+                ):
+                    return MatchTargetsResult(False, MatchFailedReason.ASSISTANT)
+
```

Comment:
I don't like this logic, and all the ones following it.

If some unexposed entity happens to match required features, it does not mean that the problem is that it is unexposed. You would first need to establish that it has a matching name in the first place.

So what you want to do instead is to have a second set of candidates, consisting only of unexposed entities and do the same operations on them as on normal candidates. If at some point normal candidates don't match and unexposed do match, then that is where you return `MatchFailedReason.ASSISTANT`.